### PR TITLE
Adding clouds

### PR DESCRIPTION
**Pull request by AvitalFriedland**

### CODE REVIEW

File: azure/Kqlmagic/kql_engine.py [hunk 1]
```diff
@@ -151,7 +151,7 @@ def validate(self, **options):
     }
     _SECRET_KEYS = {ConnStrKeys.CLIENTSECRET, ConnStrKeys.APPKEY, ConnStrKeys.PASSWORD, ConnStrKeys.CERTIFICATE_THUMBPRINT}
     _NOT_INHERITABLE_KEYS = {ConnStrKeys.APPKEY, ConnStrKeys.ALIAS}
-    _OPTIONAL_KEYS = {ConnStrKeys.TENANT, ConnStrKeys.ALIAS}
+    _OPTIONAL_KEYS = {ConnStrKeys.TENANT, ConnStrKeys.ALIAS, ConnStrKeys.CLIENTID}
```
Comment (mbnshtck) [inline azure/Kqlmagic/kql_engine.py]: I am not sure this is correct? as for some of the combinations clientId is not optional

File: azure/Kqlmagic/kql_magic.py [hunk 2]
```diff
@@ -194,6 +197,19 @@ def validate_cloud(self, cloud):
             )
 
 
+    @validate("login_code_destination")
+    def _valid_value_login_code_destination(self, proposal):
+        dest = proposal["value"]
```
Comment (mbnshtck) [inline azure/Kqlmagic/kql_magic.py]: it is better to make case insensitive

File: azure/Kqlmagic/kql_magic.py [hunk 3]
```diff
@@ -194,6 +197,19 @@ def validate_cloud(self, cloud):
...
+        if dest != "browser":
+            if not self.check_mailto(dest):
+                message = "The 'login_code_destination' trait of a {0} instance {1}".format(Constants.MAGIC_CLASS_NAME, str(e))
```
Comment (mbnshtck) [inline azure/Kqlmagic/kql_magic.py]: e is not set. the code you copy paste is a try/catch and the catch containe the exception in e. you can do the same if check_mailto will throw exception when it fails

File: azure/Kqlmagic/kql_magic.py [hunk 4]
```diff
@@ -194,6 +197,19 @@ def validate_cloud(self, cloud):
...
+                raise TraitError(message)
+        return proposal["value"]
+
+    def check_mailto(self, dest):
+        if dest.find("@")!= -1:
```
Comment (mbnshtck) [inline azure/Kqlmagic/kql_magic.py]: this is too minimal check, please google for a regular expression that checks  email pattern and use it.

File: azure/Kqlmagic/kql_magic.py [hunk 5]
```diff
@@ -194,6 +197,19 @@ def validate_cloud(self, cloud):
...
 
 
+    @validate("login_code_destination")
+    def _valid_value_login_code_destination(self, proposal):
+        dest = proposal["value"]
+        if dest != "browser":
+            if not self.check_mailto(dest):
+                message = "The 'login_code_destination' trait of a {0} instance {1}".format(Constants.MAGIC_CLASS_NAME, str(e))
+                raise TraitError(message)
+        return proposal["value"]
```
Comment (mbnshtck) [inline azure/Kqlmagic/kql_magic.py]: not suer it can be done, but I suggest you return a lowercase of proposal["value"] to make it case incensitive

File: azure/Kqlmagic/my_aad_helper.py [hunk 6]
```diff
@@ -138,6 +138,12 @@ def acquire_token(self, **options):
             url = code[OAuth2DeviceCodeResponseParameters.VERIFICATION_URL]
             device_code = code[OAuth2DeviceCodeResponseParameters.USER_CODE].strip()
             
+
+            if options.get("login_code_destination") !="browser" or  options.get("notebook_app")=="papermill":
```
Comment (mbnshtck) [inline azure/Kqlmagic/my_aad_helper.py]: the or  options.get("notebook_app")=="papermill" won't work, because it doesnot guarantee the or  login_code_destination is set to an email. Maybe the right thing to do, is if it is papermill and login_code_destination is set to browser, to return an error.

File: azure/Kqlmagic/my_aad_helper.py [hunk 7]
```diff
@@ -218,5 +224,21 @@ def acquire_token(self, **options):
             raise AuthenticationError("Unknown authentication method.")
         return self._get_header(token)
 
+
+    def send_email(self, message, mailto):
+
+        port = 465  # For SSL
+        smtp_server = "smtp.gmail.com"
+        sender_email = "dev.kql.test@gmail.com"  # Enter your address
```
Comment (mbnshtck) [inline azure/Kqlmagic/my_aad_helper.py]: please make email to start with kqlmagic@gamil.com

File: azure/Kqlmagic/my_aad_helper.py [hunk 8]
```diff
@@ -218,5 +224,21 @@ def acquire_token(self, **options):
...
+
+        receiver_email = mailto # Enter receiver address
+
+        password = "Kc0qpELOz8V3"
```
Comment (mbnshtck) [inline azure/Kqlmagic/my_aad_helper.py]: is it a permanent password? If it isn't and it must be modified every X days, how we maintain it? anyhow it should be the password of Kqlmagic. To have the password in the code is problemtic, (but I have no clear idea how to avoid it.

File: azure/Kqlmagic/my_aad_helper.py [hunk 9]
```diff
@@ -218,5 +224,21 @@ def acquire_token(self, **options):
             raise AuthenticationError("Unknown authentication method.")
         return self._get_header(token)
 
+
+    def send_email(self, message, mailto):
+
+        port = 465  # For SSL
+        smtp_server = "smtp.gmail.com"
```
Comment (mbnshtck) [inline azure/Kqlmagic/my_aad_helper.py]: smtp server should be microsoft service

File: azure/Kqlmagic/kql_engine.py (hunk 1 above)
Comment (AvitalFriedland) [reply azure/Kqlmagic/kql_engine.py]: should I delete it entirely here?

File: azure/Kqlmagic/kql_magic.py (hunk 3 above)
Comment (AvitalFriedland) [reply azure/Kqlmagic/kql_magic.py]: yes I did that this morning..somehow got lost. fixed it now

File: azure/Kqlmagic/my_aad_helper.py [hunk 10]
```diff
@@ -62,33 +62,50 @@ class AuthenticationMethod(Enum):
...
+        "mooncake": "https://login.partner.microsoftonline.cn",
+        "fairfax": "https://login.microsoftonline.us",
+        "blackforest": "https://login.microsoftonline.de",
+}
+
+
 class _MyAadHelper(object):
-    def __init__(self, kcsb, default_clientid):
+    def __init__(self, kcsb, default_clientid, cloud):
+        cloud = cloud or "public"
```
Comment (mbnshtck) [inline azure/Kqlmagic/my_aad_helper.py]: instead of passing cloud pass **options

File: azure/Kqlmagic/my_aad_helper.py [hunk 11]
```diff
@@ -62,33 +62,50 @@ class AuthenticationMethod(Enum):
...
+        if cloud.find("://") >= 0:
+            cloud_url = cloud
+        else:
+            cloud_url = _CLOUD_AAD_URLS.get(cloud)
+
+
```
Comment (mbnshtck) [inline azure/Kqlmagic/my_aad_helper.py]: please rename cloud_url to aad_login_url

File: azure/Kqlmagic/my_aad_helper.py [hunk 12]
```diff
@@ -120,58 +137,64 @@ def acquire_token(self, **options):
...
-                </script>
-
-                </body></html>"""
-            )
-
-            if options.get("notebook_app") in ["visualstudiocode", "ipython"]:
-                Display.show_window('verification_url', url, **options)
-                # Display.showInfoMessage("Code: {0}".format(device_code))
-                Display.showInfoMessage("Copy code: {0} to verification url: {1} and authenticate".format(device_code, url), **options)
+            if options.get("login_code_destination") !="browser" or  options.get("notebook_app")=="papermill":
```
Comment (mbnshtck) [inline azure/Kqlmagic/my_aad_helper.py]: this is papermill should not be part of this PR

File: azure/Kqlmagic/my_aad_helper.py [hunk 13]
```diff
@@ -201,5 +224,21 @@ def acquire_token(self, **options):
             raise AuthenticationError("Unknown authentication method.")
         return self._get_header(token)
 
+
+    def send_email(self, message, mailto):
+
+        port = 465  # For SSL
+        smtp_server = "smtp.gmail.com"
```
Comment (mbnshtck) [inline azure/Kqlmagic/my_aad_helper.py]: should not be here, part of papermill PR

File: azure/Kqlmagic/draft_client.py [hunk 14]
```diff
@@ -41,12 +41,27 @@ class DraftClient(object):
     _API_VERSION = "v1"
     _GET_SCHEMA_QUERY = ".show schema"
 
-    def __init__(self, conn_kv: dict, domain: str, data_source: str):
+    _CLOUD_AAD_URLS={
```
[the remark this answers is not in the document]
Comment (AvitalFriedland) [reply azure/Kqlmagic/draft_client.py]: what do you mean by "In case it is not the original public url, just leave it as is without modifying it"? in kusto client we dont support custom cloud urls

File: azure/Kqlmagic/la_engine.py [hunk 15]
```diff
@@ -36,9 +36,9 @@ class LoganalyticsEngine(KqlEngine):
...
     # ----------------
 
-    def __init__(self, conn_str, user_ns: dict, current=None, cloud=None):
+    def __init__(self, conn_str, user_ns: dict, current=None, **options):
         super().__init__()
         self._parsed_conn = self._parse_common_connection_str(
             conn_str, current, self._URI_SCHEMA_NAME, self._MANDATORY_KEY, self._VALID_KEYS_COMBINATIONS, user_ns
         )
-        self.client = DraftClient(self._parsed_conn, self._DOMAIN, self._DATA_SOURCE, cloud)
+        self.client = DraftClient(self._parsed_conn, self._DOMAIN, self._DATA_SOURCE, **options)
```
Comment (mbnshtck) [inline azure/Kqlmagic/la_engine.py]: you forgot to correct _DATA_SOURCE = "https://api.loganalytics.{0}" (line 19)

File: azure/Kqlmagic/connection.py [hunk 16]
```diff
@@ -52,14 +52,14 @@ def __init__(self, connect_str, user_ns:dict, **options):
         last_current = self.last_current_by_engine.get(engine.__name__)
 
         if engine != KustoEngine:
-            conn_engine = engine(connect_str, user_ns, last_current,options.get("cloud") )
+            conn_engine = engine(connect_str, user_ns, last_current,**options)
```
Comment (mbnshtck) [inline azure/Kqlmagic/connection.py]: just to be consistent, make the calll with parameter order as in line62

File: azure/Kqlmagic/kusto_client.py [hunk 17]
```diff
@@ -80,19 +94,47 @@ def __init__(self, conn_kv:dict):
...
             In case your tenant is not microsoft please use this param.
         """
+        cloud = options.get("cloud")
 
         cluster_name = conn_kv[ConnStrKeys.CLUSTER]
-        data_source = cluster_name if cluster_name.find("://") >= 0 else self._DATA_SOURCE_TEMPLATE.format(cluster_name)
+
+        if cluster_name.find("://") >= 0:
+            data_source = cluster_name
+            cloud = cloud or self.getCloudFromHTTP(data_source)
```
Comment (mbnshtck) [inline azure/Kqlmagic/kusto_client.py]: Here cloud will always have value and can't be none, so there is no point to try to retreive it from the data_source. So just remove this line. We won't try to derive it from his data_source url, he will have to set the cloud option properly.

File: azure/Kqlmagic/kusto_client.py [hunk 18]
```diff
@@ -80,19 +94,47 @@ def __init__(self, conn_kv:dict):
...
+
+        else:
+            cloud = cloud or "public"
```
Comment (mbnshtck) [inline azure/Kqlmagic/kusto_client.py]: again cloud can't be None, so just remove this line

File: azure/Kqlmagic/kusto_client.py [hunk 19]
```diff
@@ -80,19 +94,47 @@ def __init__(self, conn_kv:dict):
...
-
-        self._aad_helper = _MyAadHelper(ConnKeysKCSB(conn_kv, auth_resource), self._DEFAULT_CLIENTID) if conn_kv.get(ConnStrKeys.ANONYMOUS) is None else None
+            if cloud.find("://") >=0:
+                raise Exception("error: url for cloud not supported")
+            auth_resource = "https://kusto.kusto.{0}".format(self._CLOUD_URLS.get(cloud))
+
+        self._aad_helper = _MyAadHelper(ConnKeysKCSB(conn_kv, auth_resource), self._DEFAULT_CLIENTID, **options) if conn_kv.get(ConnStrKeys.ANONYMOUS) is None else None
+
+
+    def getCloudFromHTTP(self, http):
```
Comment (mbnshtck) [inline azure/Kqlmagic/kusto_client.py]: this function is currently not needed. But don't remove it, as we can later to add one more cloud option: 'auto' which will mean that the cloud will be derived from the url (in this case if a cluster is specified which is not a full url, we will assume public)

File: azure/Kqlmagic/draft_client.py [hunk 20]
```diff
@@ -41,12 +41,30 @@ class DraftClient(object):
...
+    "public": "https://api.{0}.io",
+    "mooncake":"https://api.{0}.azure.cn",
+    "fairfax":"https://api.{0}.us",
+    "blackforest":"https://api.{0}.de",
+}
```
Comment (mbnshtck) [inline azure/Kqlmagic/draft_client.py]: this assumes that applicationInsights and loganalytics will always have the same pattern. I would like not to assume it. and do a if (https://api.loganalytics.io) then map1, if ((https://api.applicationinsights.io)) then map2 else dont-map leave as is

File: azure/Kqlmagic/draft_client.py [hunk 21]
```diff
@@ -41,12 +41,30 @@ class DraftClient(object):
...
+
+    def __init__(self, conn_kv: dict, domain: str, data_source: str, **options):
+
+        cloud = options.get("cloud") or "public"
         self._domain = domain
-        self._data_source = data_source
+        if data_source.find("loganalytics") >= 0:
+            self._data_source = self._CLOUD_AAD_URLS.get(cloud).format("loganalytics")
+        elif data_source.find("applicationinsights") >= 0:
+            self._data_source = self._CLOUD_AAD_URLS.get(cloud).format("applicationinsights")
```
Comment (mbnshtck) [inline azure/Kqlmagic/draft_client.py]: the engine should pass the original public url, see the comment above on _CLOUD_AAD_URLS

File: azure/Kqlmagic/my_aad_helper.py [hunk 22]
```diff
@@ -120,58 +139,69 @@ def acquire_token(self, **options):
...
-
-            if options.get("notebook_app") in ["visualstudiocode", "ipython"]:
-                Display.show_window('verification_url', url, **options)
-                # Display.showInfoMessage("Code: {0}".format(device_code))
-                Display.showInfoMessage("Copy code: {0} to verification url: {1} and authenticate".format(device_code, url), **options)
+                kv = Parser.parse_and_get_kv_string(options.get('code_notification_email'), {})
+                
+                self.send_email(email_message, kv)
+               
             else:
```
Comment (mbnshtck) [inline azure/Kqlmagic/my_aad_helper.py]: this change should not be part of this PR

File: azure/Kqlmagic/parser.py [hunk 23]
```diff
@@ -380,7 +380,16 @@ def _parse_kql_command(cls, code, user_ns: dict):
         "addkqlreftohelp": {"flag": "add_kql_ref_to_help", "readonly": "True", "config": "config.add_kql_ref_to_help"},
         "addschematohelp": {"flag": "add_schema_to_help", "readonly": "True", "config": "config.add_schema_to_help"},
         "notebookapp": {"flag": "notebook_app", "readonly": "True", "config": "config.notebook_app"},
+
+
+
         "testnotebookapp": {"flag": "test_notebook_app", "readonly": "True", "config": "config.test_notebook_app"},
+        "cloud": {"flag": "cloud", "type": "str", "config": "config.cloud"},
+
+        "logincodedestination": {"flag": "login_code_destination", "type": "str", "config": "config.login_code_destination"},
```
Comment (mbnshtck) [inline azure/Kqlmagic/parser.py]: not part of thi PR

File: azure/Kqlmagic/parser.py [hunk 24]
```diff
@@ -380,7 +380,16 @@ def _parse_kql_command(cls, code, user_ns: dict):
...
+
+        "codenotificationemail": {"flag": "code_notification_email", "readonly": "True", "config": "config.code_notification_email"},
```
Comment (mbnshtck) [inline azure/Kqlmagic/parser.py]: not part of this PR

File: azure/Kqlmagic/results.py [hunk 25]
```diff
@@ -462,38 +465,61 @@ def show_chart(self, **kwargs):
...
         elif c.get("fig"):
-            if not ipywidgets_installed or Display.notebooks_host or options.get("notebook_app") in ["jupyterlab", "visualstudiocode", "ipython"]:
+            if options.get("notebook_app")=="papermill" or not ipywidgets_installed or Display.notebooks_host or options.get("notebook_app") in ["jupyterlab", "visualstudiocode", "ipython"]:
+                if options.get("notebook_app")=="papermill":
+                    x = self.to_image(**options)
+                    x1= base64.b64encode(x)
+                    s1 = x1.decode("utf-8")
+
+                    self.display_image(s1)
+                    return
```
Comment (mbnshtck) [inline azure/Kqlmagic/results.py]: not part of this PR

File: azure/Kqlmagic/results.py [hunk 26]
```diff
@@ -15,6 +15,9 @@
 import uuid
 import prettytable
 
+import base64
+from IPython.display import Image
```
Comment (mbnshtck) [inline azure/Kqlmagic/results.py]: not part of this PR

File: azure/Kqlmagic/kql_magic.py [hunk 27]
```diff
@@ -167,6 +181,43 @@ class Kqlmagic(Magics, Configurable):
...
+
+    def validate_cloud(self, cloud):
+        valid_set =  {"public","mooncake","fairfax","blackforest","usnat","ussec"}
+        if not (cloud.find("://")>=0 or cloud.lower() in valid_set):
+            raise ValueError(
+                "must be a known cloud name or custom URL, but a value of {0} was specified.".format(cloud)
+            )
+
+
+    @validate("login_code_destination")
```
Comment (mbnshtck) [inline azure/Kqlmagic/kql_magic.py]: not part of this PR

File: azure/Kqlmagic/kql_magic.py [hunk 28]
```diff
@@ -167,6 +181,43 @@ class Kqlmagic(Magics, Configurable):
...
+    def _valid_value_login_code_destination(self, proposal):
```
Comment (mbnshtck) [inline azure/Kqlmagic/kql_magic.py]: not part of this PR

File: azure/Kqlmagic/kql_magic.py [hunk 29]
```diff
@@ -848,10 +899,15 @@ def _override_default_configuration(ip, load_mode):
     if app is not None:
         lookup_key = app.lower().strip().strip("\"'").replace("_", "").replace("-", "").replace("/", "")
         app = {"jupyterlab": "jupyterlab", "jupyternotebook": "jupyternotebook", "ipython": "ipython", "visualstudiocode": "visualstudiocode", 
-                "lab": "jupyterlab", "notebook": "jupyternotebook", "ipy": "ipython", "vsc": "visualstudiocode"}.get(lookup_key)
+                "lab": "jupyterlab", "notebook": "jupyternotebook", "ipy": "ipython", "vsc": "visualstudiocode", "papermill":"papermill"}.get(lookup_key)
```
Comment (mbnshtck) [inline azure/Kqlmagic/kql_magic.py]: not part of this PR

File: azure/Kqlmagic/kql_magic.py [hunk 30]
```diff
@@ -848,10 +899,15 @@ def _override_default_configuration(ip, load_mode):
...
         if app is not None:
             ip.run_line_magic("config", '{0}.notebook_app = "{1}"'.format(Constants.MAGIC_CLASS_NAME, app.strip()))
 
+    email_config = os.getenv("{0}_CODE_NOTIFICATION_EMAIL".format(Constants.MAGIC_CLASS_NAME.upper()))
```
Comment (mbnshtck) [inline azure/Kqlmagic/kql_magic.py]: not part of this PR

File: azure/Kqlmagic/draft_client.py (hunk 20 above)
Comment (AvitalFriedland) [reply azure/Kqlmagic/draft_client.py]: and make the same maps identical for now?

File: azure/Kqlmagic/draft_client.py (hunk 21 above)
Comment (AvitalFriedland) [reply azure/Kqlmagic/draft_client.py]: so there is no need to update data source in draft client?

File: azure/Kqlmagic/my_aad_helper.py [hunk 31]
```diff
@@ -62,33 +66,49 @@ class AuthenticationMethod(Enum):
...
+
+
 class _MyAadHelper(object):
-    def __init__(self, kcsb, default_clientid):
+    def __init__(self, kcsb, default_clientid, **options):
+        cloud = options.get("cloud") or "public"
+        if kcsb.conn_kv.get("aadurl"):
+            aad_login_url = kcsb.conn_kv.get("aadurl")
+        else:
+            aad_login_url = _CLOUD_AAD_URLS.get(cloud)
```
Comment (mbnshtck) [inline azure/Kqlmagic/my_aad_helper.py]: pls raise exception in case cloud not found in list

File: azure/Kqlmagic/my_aad_helper.py [hunk 32]
```diff
@@ -62,33 +66,49 @@ class AuthenticationMethod(Enum):
...
+        "fairfax": "https://login.microsoftonline.us",
+        "blackforest": "https://login.microsoftonline.de",
+}
+
+
 class _MyAadHelper(object):
-    def __init__(self, kcsb, default_clientid):
+    def __init__(self, kcsb, default_clientid, **options):
+        cloud = options.get("cloud") or "public"
+        if kcsb.conn_kv.get("aadurl"):
```
Comment (mbnshtck) [inline azure/Kqlmagic/my_aad_helper.py]: pls use ConnStrKeys.AAD_URL instead of "aadurl"

File: azure/Kqlmagic/parser.py [hunk 33]
```diff
@@ -380,7 +380,16 @@ def _parse_kql_command(cls, code, user_ns: dict):
         "addkqlreftohelp": {"flag": "add_kql_ref_to_help", "readonly": "True", "config": "config.add_kql_ref_to_help"},
         "addschematohelp": {"flag": "add_schema_to_help", "readonly": "True", "config": "config.add_schema_to_help"},
         "notebookapp": {"flag": "notebook_app", "readonly": "True", "config": "config.notebook_app"},
+
+
+
         "testnotebookapp": {"flag": "test_notebook_app", "readonly": "True", "config": "config.test_notebook_app"},
+        "cloud": {"flag": "cloud", "type": "str", "config": "config.cloud"},
+
+        # "logincodedestination": {"flag": "login_code_destination", "type": "str", "config": "config.login_code_destination"},
```
Comment (mbnshtck) [inline azure/Kqlmagic/parser.py]: not part of this PR

File: azure/Kqlmagic/draft_client.py [hunk 34]
```diff
@@ -41,14 +41,45 @@ class DraftClient(object):
...
+            logger().debug("draft_client.py :: __init__ :  self._data_source from _CLOUD_AAD_URLS: {0}".format(self._data_source))
+
+        elif data_source.find("applicationinsights") >= 0:
+            self._data_source = self._CLOUD_AAD_URLS.get(cloud).format("applicationinsights")  
+            logger().debug("draft_client.py :: __init__ :  self._data_source from _CLOUD_AAD_URLS: {0}".format(self._data_source))
+
+
+        logger().debug("draft_client.py :: __init__ :  conn_kv[\"datasourceurl\"]: {0}".format(conn_kv.get("datasourceurl")))
+
+        if conn_kv.get("datasourceurl"):
```
Comment (mbnshtck) [inline azure/Kqlmagic/draft_client.py]: the code above us misleading, there should be at the begining if conn_kv.get("datasourceurl")... elsif..

File: azure/Kqlmagic/kusto_client.py [hunk 35]
```diff
@@ -80,19 +94,47 @@ def __init__(self, conn_kv:dict):
...
 
         self._mgmt_endpoint = self._MGMT_ENDPOINT_TEMPLATE.format(data_source, self._MGMT_ENDPOINT_VERSION)
         self._query_endpoint = self._QUERY_ENDPOINT_TEMPLATE.format(data_source, self._QUERY_ENDPOINT_VERSION)
-        _FQN_DRAFT_PROXY_CLUSTER_PATTERN = re.compile(r"http(s?)\:\/\/ade\.(int\.)?(applicationinsights|loganalytics)\.io.*$")
+        _FQN_DRAFT_PROXY_CLUSTER_PATTERN = re.compile(r"http(s?)\:\/\/ade\.(int\.)?(applicationinsights|loganalytics)\.(io|cn|us|de).*$")
         auth_resource = data_source
 
         if _FQN_DRAFT_PROXY_CLUSTER_PATTERN.match(data_source):
-            auth_resource = "https://kusto.kusto.windows.net"
-
```
Comment (mbnshtck) [inline azure/Kqlmagic/kusto_client.py]: can be removed
